text/0000-type-ascription.md

```diff
@@ -0,0 +1,175 @@
+annotating a sub-expression or sub-pattern with a type. Currently, this is only
+possible by extracting the sub-expression into a variable using a `let`
+statement and/or giving a type for a whole expression or pattern. This is un-
+ergonomic, and sometimes impossible due to lifetime issues. Specifically, a
+variable has lifetime of its enclosing scope, but a sub-expression's lifetime is
+typically limited to the nearest semi-colon.
+
+Typical use cases are where a function's return type is generic (e.g., collect)
+and where we want to force a coercion.
+
+Type ascription can also be used for documentation and debugging - where it is
+unclear from the code which type will be inferred, type ascription can be used
+to precisely communicate expectations to the compiler or other programmers.
+
+By allowing type ascription in more places, we remove the inconsistency that
+type ascription is currently only allowed on top-level patterns.
+
+## Examples:
+
+Generic return type:
+
+```
+// Current.
+let z = if ... {
+    let x: Vec<_> = foo.enumerate().collect();
+    x
+} else {
+    ...
+};
+
+// With type ascription.
+let z = if ... {
+    foo.enumerate().collect(): Vec<_>
+} else {
+    ...
+};
+```
+
+Coercion:
+
+```
+fn foo<T>(a: T, b: T) { ... }
+
+// Current.
+let x = [1u32, 2, 4];
+let y = [3u32];
+...
+let x: &[_] = &x;
+let y: &[_] = &y;
+foo(x, y);
+
+// With type ascription.
+let x = [1u32, 2, 4];
+let y = [3u32];
+...
+foo(x: &[_], y: &[_]);
+```
+
+In patterns:
+
+```
+struct Foo<T> { a: T, b: String }
+
+// Current
+fn foo(Foo { a, .. }: Foo<i32>) { ... }
+
+// With type ascription.
+fn foo(Foo { a: i32, .. }) { ... }
+```
+
+
+# Detailed design
+
+The syntax of expressions is extended with type ascription:
+
+```
+e ::= ... | e: T
+```
+
+where `e` is an expression and `T` is a type. Type ascription has the same
+precedence as explicit coercions using `as`.
+
+When type checking `e: T`, `e` must have type `T`. The `must have type` test
+includes implicit coercions and subtyping, but not explicit coercions. `T` may
+be any well-formed type.
+
+At runtime, type ascription is a no-op, unless an implicit coercion was used in
+type checking, in which case the dynamic semantics of a type ascription
+expression are exactly those of the implicit coercion.
+
+The syntax of sub-patterns is extended to include an optional type ascription.
+Old syntax:
+
+```
+P ::= SP: T | SP
+SP ::= var | 'box' SP | ...
+```
+
+where `P` is a pattern, `SP` is a sub-pattern, `T` is a type, and `var` is a
+variable name.
+
+New syntax:
+
+```
+P ::= SP: T | SP
+SP ::= var | 'box' P | ...
+```
+
+Type ascription in patterns has the narrowest precedence, e.g., `box x: T` means
+`box (x: T)`.
+
+In type checking, if an expression is matched against a pattern, when matching
+a sub-pattern the matching sub-expression must have the ascribed type (again,
+this check includes subtyping and implicit coercion). Types in patterns play no
+role at runtime.
```

Comment:
How does type ascription in patterns interact with `let` and function declarations? Presumably, `let` would now just be followed by a pattern (rather than a pattern optionally followed by a colon then a type), and the same for function parameters. Function parameters are required to have their types fully specified, but type ascription complicates this in that `fn foo((x: i32, y: i64))` should be valid as well as the current `fn foo((x, y): (i32, i64))`. Presumably the rule simply needs to be adjusted to something like *a pattern in a set of function parameters must have its type determinable without using type inference*. That rule would allow some interesting cases that don’t even need type ascription like `fn foo(NonGenericStruct(a, b), (), SingleVariantEnum)`, which may or may not be desirable.